AppCenter/AppCenter/Internals/AppDelegate/MSAppDelegateForwarder.m

```diff
@@ -446,29 +402,87 @@ - (void)custom_application:(UIApplication *)application didReceiveRemoteNotifica
 - (void)custom_application:(UIApplication *)application
     didReceiveRemoteNotification:(NSDictionary *)userInfo
           fetchCompletionHandler:(void (^)(UIBackgroundFetchResult))completionHandler {
+  __block IMP originalImp = NULL;
+  __block UIBackgroundFetchResult actualFetchResult = UIBackgroundFetchResultNoData;
+  __block short customHandlerCalledCount = 0;
+  __block short customDelegateToCallCount = 0;
+  __block MSCompletionExecutor executors = MSCompletionExecutorNone;
+
+  // This handler will be used by all the delegates, it unifies the results and execte the real handler at the end.
+  void (^commonCompletionHandler)(UIBackgroundFetchResult, MSCompletionExecutor) =
+      ^(UIBackgroundFetchResult fetchResult, MSCompletionExecutor executor) {
+
+        /*
+         * As per the Apple dicumentation:
+         * https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623013-application
+         * The `fetchCompletionHandler` is used to let the app the background time for processing the notification and
+         * download any data that will be displayed to the end users when the app will start again. There can only be
+         * one `UIBackgroundFetchResult` in the end so there is a need for triage while comparing results from
+         * delegates.
+         *
+         * Priorities are:
+         * `UIBackgroundFetchResultNewData`>`UIBackgroundFetchResultFailed`>`UIBackgroundFetchResultNoData`.
+         *  - `UIBackgroundFetchResultNewData` means at least one of the delegates did download something successfully.
+         *  - `UIBackgroundFetchResultFailed` means there was one/several downloads among the delegates but they failed.
+         *  - `UIBackgroundFetchResultNoData` means that none of the delegates had anything to download.
+         */
+        if (fetchResult == UIBackgroundFetchResultNewData || actualFetchResult == UIBackgroundFetchResultNoData) {
+          actualFetchResult = fetchResult;
+        }
+
+        // This executor is running its completion handler, remembering it.
+        executors = executors | executor;
+
+        // Count all custom executors who already ran their completion handler.
+        if (executor == MSCompletionExecutorCustom) {
+          customHandlerCalledCount++;
+        }
 
-  // Collect `UIBackgroundFetchResult` from delegates in order to call the original completion handler later.
-  __block UIBackgroundFetchResult forwardedFetchResult = UIBackgroundFetchResultNoData;
-  void (^forwardedCompletionHandler)(UIBackgroundFetchResult) = ^(UIBackgroundFetchResult fetchResult) {
-    forwardedFetchResult = fetchResult;
+        // Be sure original delegate and/or custom delegates and/or the app forwarder executed their completion handler.
+        if ((customHandlerCalledCount == customDelegateToCallCount &&
+             (executors & MSCompletionExecutorOriginal || !originalImp)) ||
+            (executor == MSCompletionExecutorForwarder)) {
```

Comment:
Let's reorder conditions, place `executor == MSCompletionExecutorForwarder` first.

Reply:
Do as you want, as soon as the logic remains the same. There are unit tests to backup this one anyways.

Reply:
I'll change.